ls1intum/Artemis
# [Refactoring] Q&A to Metis Refactoring

**[PLEASE DO NOT DEPLOY THE PR ON ANY OTHER TESTSERVER THAN TS1]**
Currently, the deployment blocks subsequents PRs on a testserver as database migrations seem to be persistent. This PR should therefore only be deployed on TS1, where is it currently tested on.



### Checklist

### Motivation and Context


This PR is the first step towards Metis (Multiplying Engagement Through Interacting Socially) - which aims to add new communication capabilities to Artemis. It **only involves refactoring** the old data model which was structured around `StudentQuestion` and `StudentQuestionAnswer`. The PR does **not add new functionality**. 

### Description

Besides renaming of properties, a major change is the introduction of a new superclass for former `StudentQuestion` and `StudentQuestionAnswer` which encapsulates the common properties.
If you are interested in details or like a more visual representation of the changes, these are the old and new data models.
<img width="969" alt="Bildschirmfoto 2021-05-29 um 21 51 42" src="https://user-images.githubusercontent.com/41366522/120083349-60812280-c0c8-11eb-857b-a062a9655132.png">

### Steps for Testing

Don't be intimidated by the number of files changed, you don't have to review them in detail 😄 . Your review should be **focused on functionality**. It is important to assure, that no regression is introduced with this refactoring PR.

Any shortcomings or errors you might find that are unrelated to this PR but concern the Q&A (and therefore future Metis) component are happily added to a list of improvements for further PRs 👍 

1. Log in to Artemis as instructor or tutor
2. Navigate to Course Management
3. Select a course
4. Navigate to the Q&A page, which serves as overview for all questions and answered in this course (also possible via  assessment dashboard, see second picture)
<img width="1417" alt="Bildschirmfoto 2021-05-29 um 21 57 35" src="https://user-images.githubusercontent.com/41366522/120083448-059bfb00-c0c9-11eb-8ca3-3b96990dce22.png">
<img width="1440" alt="Bildschirmfoto 2021-06-01 um 12 21 56" src="https://user-images.githubusercontent.com/41366522/120308112-05238000-c2d4-11eb-8162-b39145869ced.png">


Try different functionalities on this page and report any errors
6. Add questions / answers to this course (preferably as student)
7. See if the overview is updated correctly
8. Navigate to a certain exercise or lecture
9. Create / Edit / Answer questions, as tutor every added answer should be automatically be approved (green color), answers by students can be approved, edited or deleted. Questions can also be deleted

1. Log in to Artemis as students
2. Navigate to a certain exercise or lecture
Try different functionalities on this page and report any errors
3.  Create / Edit / Answer questions, as student every added answer should be un-approved by default, up-downvote other questions, edit your own questions / answers 
<img width="1411" alt="Bildschirmfoto 2021-05-29 um 22 55 27" src="https://user-images.githubusercontent.com/41366522/120084504-089ae980-c0d1-11eb-9b8e-2d63b69742e4.png">
